### PR TITLE
Add issue schema

### PR DESCRIPTION
**Pull request by nkonjeti**
## Summary

Description of PR here...

The new issue model stores the name, deadline date, release date, pitches, and type of issue (print or online). 
With the new backend endpoints, issues can be created, edited, and retrieved. 

## Changes

- Created Issue Model
- Wrote POST, PUT, GET backend endpoints

## Requests / Responses

**Request**
(Only for backend)
GET `/api/issues` Returns a list of issues

**Response**
(Only for backend)
HTTP/1.1 200 OK

```
{
   "message": "Successfully retrieved all issues.",
   "success": true,
   "result": [
       {
           "name": "testIssueUpdate",
           "deadlineDate": "2023-11-03T02:24:00.754+00:00",
           "releaseDate": "2024-11-03T02:24:00.754+00:00",
           "pitches": [],
           "_id": "618b383fb68cbf10d4518300",
           "type": "onlineIssue",
           "__v": 0
       }
   ]
}


```

**Request**
(Only for backend)
GET `/api/issues/:issueId` Get issue with ID

**Response**
(Only for backend)
HTTP/1.1 200 OK

```
{
   "success": true,
   "result": {
       "name": "testIssueUpdate",
       "deadlineDate": "2023-11-03T02:24:00.754+00:00",
       "releaseDate": "2024-11-03T02:24:00.754+00:00",
       "pitches": [],
       "_id": "618b383fb68cbf10d4518300",
       "type": "onlineIssue",
       "__v": 0
   },
   "message": "Successfully retrieved issue"
}


```

**Request**
(Only for backend)
POST `/api/issues` Create a issue

```
{
  "name": "testIssue2",
  "deadlineDate": "2023-11-03T02:24:00.754+00:00",
   "releaseDate": "2024-11-03T02:24:00.754+00:00"
   "type": "onlineIssue"
}
```

**Response**

HTTP/1.1 200 OK

```
{
   "message": "Successfully created new issue",
   "success": true,
   "result": {
       "name": "testIssue2",
       "deadlineDate": "2023-11-03T02:24:00.754+00:00",
       "releaseDate": "2024-11-03T02:24:00.754+00:00",
       "pitches": [],
       "_id": "618c98c17f69b825df18b0b8",
       "type": "onlineIssue",
       "__v": 0
   }
}



```

**Request**
(Only for backend)
PUT `/api/issues/:issueId` Update Issue with ID

```
{
  "name": "testUpdateIssue2"
}
```

**Response**
(Only for backend)
HTTP/1.1 200 OK

```
{
   "success": true,
   "message": "Successfully updated issue",
   "result": {
       "name": "testUpdateIssue2\n",
       "deadlineDate": "2023-11-03T02:24:00.754+00:00",
       "releaseDate": "2024-11-03T02:24:00.754+00:00",
       "pitches": [],
       "_id": "618c98c17f69b825df18b0b8",
       "type": "onlineIssue",
       "__v": 0
   }
}


```

### CODE REVIEW

File: api/src/models/issue.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import mongoose, { Document, Schema } from 'mongoose';
+import { IResource } from 'ssw-common';
+
+type IssueSchema = IResource & Document<any>;
+enum Type {
+  printIssue,
+  onlineIssue,
+}
```
Comment (amit-sawhney) [inline api/src/models/issue.ts:8]: See in the utils file on the api that there already exists an enum for this.

File: api/src/models/issue.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import mongoose, { Document, Schema } from 'mongoose';
+import { IResource } from 'ssw-common';
```
Comment (amit-sawhney) [inline api/src/models/issue.ts:2]: Why are we using `IResource` here?

File: common/index.d.ts [hunk 3]
```diff
@@ -118,3 +118,20 @@ export interface ITeam {
   active: boolean;
   color: string;
 }
+
+declare enum Type {
+  printIssue,
+  onlineIssue,
+}
```
Comment (amit-sawhney) [inline common/index.d.ts:125]: See how we do constants/enums in the utils file and fix this.

File: common/index.d.ts [hunk 4]
```diff
@@ -118,3 +118,20 @@ export interface ITeam {
...
+/**
+ * Interface for a Issue Schema.
+ */
+export interface IIssue {
+  _id: string;
+  name: string;
+  deadlineDate: string;
+  releaseDate: string;
+  pitches: string[];
+  type: Type;
```
Comment (amit-sawhney) [inline common/index.d.ts:136]: Should by type string

**Review by amit-sawhney — CHANGES_REQUESTED** [review]
Looks good! There are some enum/constant things I want to see fixed and then it should be good. :)

File: common/index.d.ts (hunk 4 above)
Comment (nkonjeti) [reply common/index.d.ts]: or the type would be the enum IssueFormatType right?

Comment (nkonjeti) [reply common/index.d.ts]: oh never mind!

File: api/src/models/issue.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+type IssueSchema = IIssue & Document<any>;
+/**
+ * Mongoose Schema to represent an Issue at South Side Weekly.
+ */
+const Issue = new mongoose.Schema({
+  name: { type: String, default: null, required: true },
+  deadlineDate: { type: String, default: null, required: true },
+  releaseDate: { type: String, default: null, required: true },
+  pitches: [{ type: Schema.Types.ObjectId, ref: 'Pitch' }],
+  type: { type: issueFormatEnum, required: true },
```
Comment (amit-sawhney) [inline api/src/models/issue.ts:14]: See this example.

```tsx

onboardingStatus: {
    type: String,
    enum: Object.values(onboardingStatusEnum),
    default: onboardingStatusEnum.ONBOARDING_SCHEDULED,
  },

```

**Review by amit-sawhney — APPROVED** [review]

File: api/src/api/issue.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    if (!issue) {
+      res.status(404).json({
+        success: false,
+        message: 'Issue not found with id',
+      });
+    } else {
+      res.status(200).json({
+        success: true,
+        result: issue,
+        message: `Successfully retrieved issue`,
```
Comment (AndrewLester) [inline api/src/api/issue.ts:24]:
```suggestion
        message: 'Successfully retrieved issue',
```

File: api/src/api/issue.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+);
+
+//Gets all issues
+router.get(
+  '/',
+  requireRegistered,
+  errorWrap(async (req: Request, res: Response) => {
+    const issues = await Issue.find({});
+    res.status(200).json({
+      message: `Successfully retrieved all issues.`,
```
Comment (AndrewLester) [inline api/src/api/issue.ts:37]:
```suggestion
      message: 'Successfully retrieved all issues.',
```

File: api/src/models/issue.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+type IssueSchema = IIssue & Document<any>;
+/**
+ * Mongoose Schema to represent an Issue at South Side Weekly.
+ */
+const Issue = new mongoose.Schema({
+  name: { type: String, default: null, required: true },
+  deadlineDate: { type: String, default: null, required: true },
+  releaseDate: { type: String, default: null, required: true },
+  pitches: [{ type: Schema.Types.ObjectId, ref: 'Pitch' }],
+  type: { type: String, enum: Object.values(issueFormatEnum), required: true },
```
Comment (AndrewLester) [inline api/src/models/issue.ts:14]: If the enum is called `issueFormat` should one of the two (this property called `type` or `issueFormatEnum`) be renamed to match the other?

**Review by AndrewLester — APPROVED** [review]
Looks good, just two quick changes and an open question I may be wrong about.

**Pull request merged**